PARQUET-220: Unnecessary warning in ParquetRecordReader.initialize

sircodesalotOfTheRound opened the pull request:
Rather than querying the COUNTER_METHOD up front, the counter method is resolved per object. This allows us to use the
'getCounter' method on any TaskAttemptContext with the correct signature (ignoring versions where TaskAttemptContext does
not have an appropriate method/signature - preserving current behavior).

rdblue commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ContextUtil.java:
```diff
@@ -251,17 +251,36 @@ public static Configuration getConfiguration(JobContext context) {
-    return (Counter) invoke(GET_COUNTER_METHOD, context, groupName, counterName);
+  public static Counter getCounter(TaskAttemptContext context, String groupName, String counterName) {
+    Method counterMethod = findCounterMethod(context);
+    return (Counter) invoke(counterMethod, context, groupName, counterName);
```
Is the cast still necessary now that invoke has a type parameter and casts?

sircodesalotOfTheRound replied:
Oversight, I can fix that.

rdblue commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ContextUtil.java:
```diff
@@ -251,17 +251,36 @@ public static Configuration getConfiguration(JobContext context) {
 
-  public static Counter getCounter(TaskInputOutputContext context,
-                                   String groupName, String counterName) {
-    return (Counter) invoke(GET_COUNTER_METHOD, context, groupName, counterName);
```
What happened to GET_COUNTER_METHOD? Why isn't that used in findCounterMethod?

rdblue replied:
I think this was intended to be a one-time startup cost. GET_COUNTER_METHOD was a static field that was the method. Now that this code finds them dynamically based on class it should be fine to remove GET_COUNTER_METHOD. Since this is done once per input split, I doubt the overhead is going to matter much.

sircodesalotOfTheRound replied:
I'll split the difference and cache the counter method the first time it's used. Thanks for the catch.

sircodesalotOfTheRound replied:
I replaced `GET_COUNTER_METHOD` with a cache (`COUNTER_METHODS_BY_CLASS`).

sircodesalotOfTheRound commented:
Travis build is failing with:

````
Class org.apache.parquet.hadoop.util.counters.mapreduce.MapReduceCounterLoader
 Added Method <init>, desc (Lorg/apache/hadoop/mapreduce/TaskAttemptContext;)V, access public
 Removed Method <init>, sig (Lorg/apache/hadoop/mapreduce/TaskInputOutputContext<****>;)V, desc (Lorg/apache/hadoop/mapreduce/TaskInputOutputContext;)V, access public
[WARNING] Rule 0: org.semver.enforcer.RequireSemanticVersioningConformance failed with message:
Current codebase is incompatible with version <1.7.0>. Version should be at least <2.0.0>.
````

What does this mean?

rdblue commented:
@sircodesalotOfTheRound, that means this is failing the API compatibility check. This is not binary compatible with the 1.7.0 release. I think it is safe to say that org.apach.parquet.hadoop.util.* is probably not considered public API so we just need to add an exclusion for it in the root POM. Look for the semver plugin configuration.

julienledem commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ContextUtil.java:
```diff
@@ -251,17 +260,41 @@ public static Configuration getConfiguration(JobContext context) {
+  private static <T> T invoke(Method method, Object obj, Object... args) {
     try {
-      return method.invoke(obj, args);
+      return (T)method.invoke(obj, args);
```
That would create a warning. The only safe way to do that is casting from the caller.

julienledem commented:
This change looks good to me but travis build is still failing.

sircodesalotOfTheRound commented:
Fix'd.

julienledem commented:
LGTM
@rdblue are we happy with org/apache/parquet/hadoop/util/** excluded from semver?

rdblue commented:
I think @sircodesalotOfTheRound was going to address the two comments remaining above before merging.

I'm fine excluding org/apache/parquet/hadoop/util/**

julienledem commented:
@rdblue good to go?

sahilsareen commented:
Merge this please!

julienledem commented:
@rdblue Merge?

dongjoon-hyun commented:
Any update? :)

julienledem commented:
@rdblue If that's ok, I'm going to merge this.

rdblue commented:
Yeah, sounds good.

julienledem commented:
@dongjoon-hyun @sircodesalotOfTheRound @sahilsareen merged

dongjoon-hyun commented:
Thank you so much! :)
